solver/src/naive_solver/multi_order_solver.rs

```diff
@@ -0,0 +1,369 @@
+
+use super::single_pair_settlement::{AmmSwapExactTokensForTokens, SinglePairSettlement};
+use crate::settlement::Trade;
+use anyhow::{anyhow, Result};
+use model::order::{OrderCreation, OrderKind};
+use num::{bigint::Sign, BigInt};
+use std::collections::HashMap;
+use web3::types::{Address, U256};
+
+#[derive(Default)]
+struct TokenContext {
+    address: Address,
+    reserve: BigInt,
+    buy_volume: BigInt,
+    sell_volume: BigInt,
+}
+
+impl TokenContext {
+    pub fn is_excess(&self, other: &TokenContext) -> bool {
+        &self.reserve * (&other.sell_volume - &other.buy_volume)
+            < &other.reserve * (&self.sell_volume - &self.buy_volume)
+    }
+}
+
+/**
+ * Computes a settlement using orders of a single pair and the direct AMM between those tokens.
+ * Returns an error if computation fails, orders are not already filtered for a specific token pair, or the reserve information for that
+ * pair is not available.
+ */
+pub fn solve(
+    orders: impl Iterator<Item = OrderCreation> + Clone,
+    reserves: &HashMap<Address, U256>,
+) -> Result<SinglePairSettlement> {
+    let (context_a, context_b) = split_into_contexts(orders.clone(), reserves)?;
+    let (shortage, excess) = if context_a.is_excess(&context_b) {
+        (context_b, context_a)
+    } else {
+        (context_a, context_b)
+    };
+    let uniswap_out = compute_uniswap_out(&shortage, &excess);
+    let uniswap_in = bigint_to_u256(&compute_uniswap_in(&uniswap_out, &shortage, &excess))?;
+    let uniswap_out = bigint_to_u256(&uniswap_out)?;
+    let interaction = if uniswap_out > U256::zero() {
+        Some(AmmSwapExactTokensForTokens {
+            amount_in: uniswap_in,
+            amount_out_min: uniswap_out,
+            token_in: excess.address,
+            token_out: shortage.address,
+        })
+    } else {
+        // TODO(fleupold) set correct clearing prices when supply/demand already match current uniswap price
+        None
+    };
+    // TODO(fleupold) check that all orders comply with the computed price. Otherwise, remove the least favorable excess order and try again.
+    Ok(SinglePairSettlement {
+        clearing_prices: maplit::hashmap! {
+            shortage.address => uniswap_in,
+            excess.address => uniswap_out,
+        },
+        trades: orders.into_iter().map(Trade::fully_matched).collect(),
+        interaction,
+    })
+}
+
+fn split_into_contexts(
+    orders: impl Iterator<Item = OrderCreation>,
+    reserves: &HashMap<Address, U256>,
+) -> Result<(TokenContext, TokenContext)> {
+    let mut contexts = HashMap::new();
+    for order in orders {
+        contexts.entry(order.buy_token).or_insert(TokenContext {
+            address: order.buy_token,
+            reserve: u256_to_bigint(
+                reserves
+                    .get(&order.buy_token)
+                    .ok_or_else(|| anyhow!("No reserve for token {}", &order.buy_token))?,
+            ),
+            ..Default::default()
+        });
```

Comment:
Use insert_with to avoid doing u256 conversion even entry already exists:
```suggestion
        contexts
            .entry(order.buy_token)
            .or_insert_with(|| TokenContext {
                address: order.buy_token,
                reserve: u256_to_bigint(reserves.get(&order.buy_token).expect("no reserve token")),
                ..Default::default()
            });
```
but that way can't return a result which I think we shouldn't do anyway.